<a href="https://colab.research.google.com/github/leof98/CS50/blob/main/TrabEntropia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Trabalho matemática (presenças)**

Nome: Leonardo Franco

RA: 0051352321007







 Calcular Entropia

 Calcular Entropia Máxima

H = - ∑ p(x)log(p(x))

Hmax ​= − log2​(n​)

In [3]:
import numpy as np
import pandas as pd

# H = -sum(pk * log(pk))

# Load the dataset
dataset = pd.read_csv('passwords.csv')

# Specify the column name
column_name = 'category'

# Extract the values from the specified column
column_values = dataset[column_name].dropna().values

# Calculate probabilities (p_k) for each unique value
unique_values, counts = np.unique(column_values, return_counts=True)
probs = counts / len(column_values)

# Calculate the term p_k * log2(p_k) for each unique value
term_values = probs * np.log2(probs)

# Sum up the terms to get the final entropy value
column_entropy = -np.sum(term_values)

# Calculate the maximum entropy
num_categories = len(unique_values)
max_entropy = -np.log2(1 / num_categories)

# Print intermediate steps for verification
print(f"Unique Values: {unique_values}")
print(f"Counts: {counts}")
print(f"Probabilities: {probs}")
print(f"Term Values: {term_values}\n")

# Print the final results
print(f"Entropia: {column_entropy}")
print(f"Entropia máxima: {max_entropy}")

Unique Values: ['animal' 'cool-macho' 'fluffy' 'food' 'name' 'nerdy-pop'
 'password-related' 'rebellious-rude' 'simple-alphanumeric' 'sport']
Counts: [ 29  79  44  11 183  30  15  11  61  37]
Probabilities: [0.058 0.158 0.088 0.022 0.366 0.06  0.03  0.022 0.122 0.074]
Term Values: [-0.23825259 -0.42059656 -0.30855903 -0.12113976 -0.53073091 -0.24353362
 -0.15176681 -0.12113976 -0.37027573 -0.27796849]

Entropia: 2.7839632564166417
Entropia máxima: 3.321928094887362


Dataset utilizado: https://www.kaggle.com/datasets/sujaykapadnis/bad-passwords-never-use-them


Coluna categórica escolhida: 'Category'

Entropia ≈ 2.7839

Entropia máxima ≈ 3.3219